In [1]:
import random

In [2]:
def average_py(n):
    s = 0
    for i in range(n):
        s += random.random()
    return s / n 

In [3]:
n = 10000000

In [4]:
%time average_py(n)

CPU times: total: 1.77 s
Wall time: 1.8 s


0.5000673231997611

In [5]:
%timeit average_py(n)

2.8 s ± 253 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [6]:
%time sum([random.random() for _ in range(n)]) / n 

CPU times: total: 3.42 s
Wall time: 3.84 s


0.4999330762581073

<span style='font-size:17px'><b>Basic Python Benchmarking</b>

* <b>Function Definition:</b> Creates a pure Python function 'average_py' that uses a simple 'for' loop to calculate the mean of $n$ random numbers.
* <b>Basline Performance:</b> Benchmarks the function with 10 million iterations, recording a wall time of 1.31 seconds for a single execution.
* <b>Statistical Timing:</b> Uses '%timeit' to provide a more robust average, showing approximately 1.36 seconds per loop across 7 different runs.
* <b>List Comprehension Comparison:</b> Tests a one-liner approach using 'sum()' on a list comprehension, which performs slower at 2.17 seconds due to memory overhead.

In [7]:
import numpy as np

In [8]:
def average_np(n):
    s = np.random.random(n)
    return s.mean()


In [9]:
%time average_np(n)

CPU times: total: 281 ms
Wall time: 317 ms


np.float64(0.5000940890445686)

In [10]:
%timeit average_np(n)

338 ms ± 43.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [11]:
s = np.random.random(n)
s.nbytes

80000000

<span style='font-size:17px'><b>NumPy Vectorization</b>

* <b>Function Optimization:</b> Implements 'average_np' using NumPy's 'random' and 'mean' methods to replace the manual Python loop.
* <b>Syntax Debugging:</b> Identifies that 's.mean' returns the function object whereas 's.mean()' is required to execute the calculation.
* <b>Performance Leap:</b> Acheives a wall time of 133 ms for 10 million numbers, making it roughly 10x faster than the pure Python loop.
* <b>Vectorized Execution:</b> Demonstrates how NumPy processes the entire arra at once in C, drastically reducing the overhead compared to Python iteration.

In [12]:
import numba

In [13]:
average_nb = numba.jit(average_py)

In [14]:
%time average_nb(n)

CPU times: total: 2.27 s
Wall time: 3.61 s


0.49991310307299747

In [15]:
%time average_nb(n)

CPU times: total: 188 ms
Wall time: 214 ms


0.5000511505161563

In [16]:
%timeit average_nb(n)

198 ms ± 3.73 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


<span style='font-size:17px'><b>Numba Just-In-Time (JIT) Compilation</b>

* <b>Compilation Overhead:</b> Incurs a one-time 'warm-up' cost during the first execution as Numba translates Python bytecode into optimized machine code.
* <b>Execution Efficiency:</b> Reduces execution time from ~1.3 seconds to 119 milliseconds once the function is cached in its compiled state.
* <b>Seamless Integration:</b> Demonstrates the 'numba.jit()' wrapper's ability to take a standard Python function and optimize it without manual code rewrites.
* <b>Performance Parity:</b> Acheives results comparable to NumPy's vectorized methods by running calculations directly at the CPU instruction level.

!pip install Cython

In [17]:
%load_ext Cython

In [18]:
%%cython -a 
import random
def average_cy1(int n):
    cdef int i
    cdef float s = 0 
    for i in range(n):
        s += random.random()
    return s/n

In [19]:
%time average_cy1(n)

CPU times: total: 2.86 s
Wall time: 3.11 s


0.5000099539756775

In [20]:
%timeit average_cy1(n)

3.09 s ± 129 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [21]:
%%cython 
from libc.stdlib cimport rand 
cdef extern from 'limits.h':
    int INT_MAX
cdef int i
cdef float rn
for i in range(5):
    rn = rand() / INT_MAX
    print(rn)

1.2555159628391266e-05
1.1391937732696533e-05
2.6565976440906525e-06
1.3106036931276321e-05
1.0841060429811478e-05


In [22]:
%%cython -a 
from libc.stdlib cimport rand
cdef extern from 'limits.h':
    int INT_MAX
def average_cy2(int n):
    cdef int i 
    cdef float s = 0
    for i in range(n):
        s += rand() / INT_MAX
    return s / n

In [23]:
%time average_cy2(n)

CPU times: total: 312 ms
Wall time: 389 ms


7.628993898833869e-06

In [24]:
%timeit average_cy2(n)

257 ms ± 12.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


<span style='font-size:17px'><b>Cython and C-Library Integration</b>

* <b>Library Transition:</b> Shifts from the high-level Python 'random' module to the low-level 'libc.stdlib' C library to eliminate interpreter overhead.
* <b>Static Declaration:</b> Uses 'cdef' to explicitly define variable types and external C constants like 'INT_MAX', allowing for direct machine-code mapping.
* <b>Optimization Analysis:</b> Employs the '-a' annotation flag to identify and remove 'yellow' Python-interaction lines, achieving a 'white' pure-C execution loop.
* <b>Efficiency Breakthrough:</b> Reduces execution time from ~2.0 seconds to 153 milliseconds by bypassing the Python virtual machine entirely during the calculation.

In [25]:
def is_prime(I):
    if I % 2 == 0: return False
    for i in range(3, int(I ** 0.5) + 1, 2):
        if I % i == 0: return False
    return True


In [26]:
n = int(1e8 + 3)
n

100000003

In [27]:
%time is_prime(n)

CPU times: total: 0 ns
Wall time: 68.4 μs


False

In [28]:
p1 = int(1e8 + 7)
p1

100000007

In [29]:
%time is_prime(p1)

CPU times: total: 0 ns
Wall time: 576 μs


True

In [30]:
p2 = 100109100129162907

In [31]:
p2.bit_length()

57

In [61]:
%time is_prime(p2)

CPU times: total: 14.2 s
Wall time: 14.6 s


True

<span style='font-size:17px'><b>Computational Primality Testing & Efficiency</b>

* <b>Optimized Trial Division:</b> The 'is_prime' function implements a loop that only checks odd numbers up to the square root of the target ($\sqrt{1}$), which is a significantly more efficient approach than checking every number up to $I$.
* <b>Time Complexity Practice:</b> The results show a massive jump in 'Wall Time' as the numbers grow; while $10^8 + 3$ is processed in **$68.4 \ \ mu s$**, the 57-bit prime $p2$ requires **$15$ seconds** to verify.
* <b>Bit-Length Impact:</b> The use of 'p2.bit_length()' highlights that asthe binary size of a number increases (to **57 bits** in this case), the number of potential divisors grows, causing a linear algorithm to slow down noticeably.
* <b>Performance Monitoring:</b> By using the '%time' magic command, the output distinguishes **CPU time** (active processing) and **Wall time** (total time elapsed), which is identifying bottlenecks in cryptographic or mathematical scripts.

In [33]:
is_prime_nb = numba.jit(is_prime)

In [34]:
%time is_prime_nb(n)

CPU times: total: 156 ms
Wall time: 192 ms


False

In [37]:
%time is_prime_nb(n)

CPU times: total: 0 ns
Wall time: 11.2 μs


False

In [38]:
%time is_prime_nb(p1)

CPU times: total: 0 ns
Wall time: 36.5 μs


True

In [39]:
%time is_prime_nb(p2)

CPU times: total: 641 ms
Wall time: 657 ms


True

In [54]:
%%cython
def is_prime_cy1(I):
    if I % 2 == 0: return False
    for i in range(3, int(I ** 0.5) + 1, 2):
        if I % i == 0: return False
    return True

Content of stdout:
_cython_magic_93ea45f039ee42cfd31d72bd8d8f006184105d0586ed86141cda43136320a76a.c
   Creating library C:\Users\wyatt\.ipython\cython\_cython_magic_93ea45f039ee42cfd31d72bd8d8f006184105d0586ed86141cda43136320a76a.cp313-win_amd64.lib and object C:\Users\wyatt\.ipython\cython\_cython_magic_93ea45f039ee42cfd31d72bd8d8f006184105d0586ed86141cda43136320a76a.cp313-win_amd64.exp
Generating code
Finished generating code

In [55]:
%timeit is_prime(p1)

464 μs ± 15.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [56]:
%timeit is_prime_cy1(p1)

334 μs ± 29.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [57]:
%%cython
def is_prime_cy2(long I):
    cdef long i
    if I % 2 == 0: return False
    for i in range(3, int(I ** 0.5) + 1, 2):
        if I % i == 0: return False
    return True

In [58]:
%timeit is_prime_cy2(p1)

14.2 μs ± 615 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [59]:
%timeit is_prime_nb(p1)

21.6 μs ± 966 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [60]:
%time is_prime_cy2(p2)

OverflowError: Python int too large to convert to C long

<span style='font-size:17px'><b>Performance Scaling and Limitations in Cython and Numba</b>

* <b>Static Typing Efficiency:</b> The transition from 'is_prime_cy1' to 'is_prime_cy2' demonstrates that explicitly declaring types (e.g., 'long I' and 'cdef long i') allows Cython to bypass Python's overhead, reducing execution time for $p1$ from **$334 \ \mu s$** results likely utilizes a 64-bit environment.
* <b>The 32-bit vs. 64-bit Bottleneck:</b> The 'OverflowError' encounterd with $p2$ occurs because the code attempts to fit a **57-bit** integer into a C 'long'; on Windows systems, this type is typically limited to 32 bits, whereas the book's successful **1.68s** result likely utilizes a 64-bit environment.
* <b>Numba's JIT advantage:</b> Numba provides a high-performance alternative to Cython with less manual configuration, successfully processing the large prime $p2$ in **$657 \ ms$** compared to the original Python time of **$15 \ seconds$**.
* <b>Compilation Latency:</b> Initial calls to Numba-decorated functions (like the first run of 'is_prime_nb') includes a one-time 'warm-up' cost for machine code compilation (**$192 \ ms$**), while subsequent calls run at near-native speeds (**$11.2 \ \mu s$**).